modules/ssl/ssl_engine_kernel.c

```diff
@@ -126,21 +126,62 @@ static int ap_array_same_str_set(apr_array_header_t *s1, apr_array_header_t *s2)
     return 1;
 }
 
-static int ssl_pk_server_compatible(modssl_pk_server_t *pks1, 
-                                    modssl_pk_server_t *pks2) 
+#ifdef HAVE_SSL_CONF_CMD
+/* Writes a hash of the ssl_ctx_param array to digest. */
+static void hash_ssl_params(apr_array_header_t *a,
+                            unsigned char digest[APR_MD5_DIGESTSIZE])
 {
-    if (!pks1 || !pks2) {
-        return 0;
+    apr_md5_ctx_t ctx;
+    int n;
+
+    apr_md5_init(&ctx);
+    for (n = 0; n < a->nelts; n++) {
+        ssl_ctx_param_t *p = &APR_ARRAY_IDX(a, n, ssl_ctx_param_t);
+
+        apr_md5_update(&ctx, p->name, strlen(p->name));
+        apr_md5_update(&ctx, "##", 2);
+        apr_md5_update(&ctx, p->value, strlen(p->value));
     }
-    /* both have the same certificates? */
-    if ((pks1->ca_name_path != pks2->ca_name_path)
-        && (!pks1->ca_name_path || !pks2->ca_name_path 
-            || strcmp(pks1->ca_name_path, pks2->ca_name_path))) {
+    apr_md5_final(digest, &ctx);
+}
+
+/* Return non-zero if the two ssl_ctx_param arrays match exactly, else
+ * zero. */
+static int array_cmp_ssl_ctx_params(apr_array_header_t *a1,
+                                    apr_array_header_t *a2)
+{
+    unsigned char dig1[APR_MD5_DIGESTSIZE], dig2[APR_MD5_DIGESTSIZE];
+
+    if (a1 == a2)
+        return 1;
+    if (a1->nelts != a2->nelts)
+        return 0;
+
+    hash_ssl_params(a1, dig1);
+    hash_ssl_params(a2, dig2);
+
+    return strcmp((char *)dig1, (char *)dig2) == 0;
```

Comment:
Are `dig1` and `dig2` null terminated strings that do not contain any nulls inside or do we need to do a `memcmp(dig1, dig2, APR_MD5_DIGESTSIZE)` instead of the `strcmp`?